### ISSUE TITLE
Spurious uninitialized variable with while loop

### ISSUE THREAD

**Issue by jez** [sorbet/sorbet#4042]
#### Input

[→ View on sorbet.run](https://sorbet.run/#%23%20typed%3A%20true%0A%0Arequire%20'sorbet-runtime'%0A%0Aextend%20T%3A%3ASig%0A%0Asig%20%7Breturns(T%3A%3AArray%5BString%5D)%7D%0Adef%20foo%0A%20%20match%20%3D%20T.unsafe(true)%0A%0A%20%20operands%20%3D%20T%3A%3AArray%5BString%5D.new%0A%0A%20%20while%20match%0A%20%20%20%20match%20%3D%20T.unsafe(false)%0A%20%20%20%20if%20!match%0A%20%20%20%20%20%20puts%20match%0A%20%20%20%20end%0A%20%20end%0A%0A%20%20operands%0Aend%0A%0Afoo)

```ruby
# typed: true

require 'sorbet-runtime'

extend T::Sig

sig {returns(T::Array[String])}
def foo
  match = T.unsafe(true)

  operands = T::Array[String].new

  while match
    match = T.unsafe(false)
    if !match
      puts match
    end
  end

  operands
end

foo
```

#### Observed output

```
editor.rb:20: Expected T::Array[String] but found T.nilable(T::Array[String]) for method result type https://srb.help/7005
    20 |  operands
          ^^^^^^^^
  Expected T::Array[String] for result type of method foo:
    editor.rb:8:
     8 |def foo
        ^^^^^^^
  Got T.nilable(T::Array[String]) originating from:
    editor.rb:11:
    11 |  operands = T::Array[String].new
                     ^^^^^^^^^^^^^^^^^^^^
    editor.rb:8: Possibly uninitialized (NilClass) in:
     8 |def foo
        ^^^^^^^
Errors: 1
```

#### Expected behavior

No errors (`operands` is initialized once, before any conditional logic, and never re-assigned, so it should never be `NilClass`).

- - -

<img width="1901" alt="Screen Shot 2021-03-01 at 5 29 58 PM" src="https://user-images.githubusercontent.com/5544532/109582880-cce18b00-7ab3-11eb-80d1-494801be4e76.png">

I think we might have another ticket for this, but I figure it'll be easier to de-dup later if we ever find it.

**Comment by neilparikh** [comment]
I think this is related to https://github.com/sorbet/sorbet/issues/96

**Comment by jez** [comment]
> I think this is related to #96

I am not yet convinced they are the same bug, so I think we should leave them both open.

**Comment by Bo98** [comment]
While you can workaround the previous example with `T.must(xs)`, things get worse when you add a rescue block:

[→ View on sorbet.run](https://sorbet.run/#%23%20typed%3A%20true%0Aextend%20T%3A%3ASig%0A%0Abegin%0A%20%20xs%20%3D%20T.let%28%5B%5D%2C%20T%3A%3AArray%5BInteger%5D%29%0A%20%20has_more%20%3D%20T.let%28true%2C%20T%3A%3ABoolean%29%0A%0A%20%20i%20%3D%200%0A%20%20while%20has_more%0A%20%20%20%20xs.concat%28%5Bi%5D%29%0A%20%20%20%20if%20i%20%3E%3D%205%0A%20%20%20%20%20%20has_more%20%3D%20false%0A%20%20%20%20end%0A%20%20end%0A%0A%20%20T.must%28xs%29.each%20do%20%7Cx%7C%0A%20%20%20%20print%20x%0A%20%20end%0Arescue%20%3D%3E%20e%0A%20%20%5B%5D%0Aend%0A)

```ruby
# typed: true
extend T::Sig

begin
  xs = T.let([], T::Array[Integer])
  has_more = T.let(true, T::Boolean)

  i = 0
  while has_more
    xs.concat([i])
    if i >= 5
      has_more = false
    end
  end

  T.must(xs).each do |x|
    print x
  end
rescue => e
  []
end
```

```
editor.rb:19: Method <keep-for-cfg> does not exist on NilClass component of T.nilable(T.class_of(<Magic>)) https://srb.help/7003
    19 |rescue => e
    20 |  []
  Got T.nilable(T.class_of(<Magic>)) originating from:
    editor.rb:2: Possibly uninitialized (NilClass) in:
     2 |extend T::Sig
     3 |
     4 |begin
     5 |  xs = T.let([], T::Array[Integer])
     6 |  has_more = T.let(true, T::Boolean)
       |...
    17 |    print x
    18 |  end
    19 |rescue => e
    20 |  []
    21 |end
  Autocorrect: Use -a to autocorrect
    editor.rb:19: Replace with T.must(rescue => e
      [])
    19 |rescue => e
    20 |  []
Errors: 1
```

AFAIK this is impossible to workaround and the autocorrect is invalid syntax.

**Comment by Bo98** [comment]
Reduced more simply to:

[→ View on sorbet.run](https://sorbet.run/#%23%20typed%3A%20true%0Aextend%20T%3A%3ASig%0A%0Abegin%0A%20%20has_more%20%3D%20T.let%28true%2C%20T%3A%3ABoolean%29%0A%20%20while%20has_more%0A%20%20%20%20has_more%20%3D%20false%0A%20%20%20%20if%20true%0A%20%20%20%20%20%20%23%20same%20error%20if%20you%20put%20%60has_more%20%3D%20false%60%20into%20here%20instead%20of%20this%0A%20%20%20%20%20%20puts%20%22any%20statement%20inside%20an%20if%20block%22%0A%20%20%20%20end%0A%20%20end%0Arescue%20%3D%3E%20e%0Aend%0A)

```ruby
# typed: true
extend T::Sig

begin
  has_more = T.let(true, T::Boolean)
  while has_more
    has_more = false
    if true
      # same error if you put `has_more = false` into here instead of this
      puts "any statement inside an if block"
    end
  end
rescue => e
end
```

```
editor.rb:13: Method <keep-for-cfg> does not exist on NilClass component of T.nilable(T.class_of(<Magic>)) https://srb.help/7003
    13 |rescue => e
        ^^^^^^^^^^^
  Got T.nilable(T.class_of(<Magic>)) originating from:
    editor.rb:2: Possibly uninitialized (NilClass) in:
     2 |extend T::Sig
     3 |
     4 |begin
     5 |  has_more = T.let(true, T::Boolean)
     6 |  while has_more
       |...
    10 |      puts "any statement inside an if block"
    11 |    end
    12 |  end
    13 |rescue => e
    14 |end
  Autocorrect: Use -a to autocorrect
    editor.rb:13: Replace with T.must(rescue => e)
    13 |rescue => e
        ^^^^^^^^^^^
Errors: 1
```

**Comment by elliottt** [comment]
Fixed in #8291
